src/ray/common/asio/postable.h

```diff
@@ -0,0 +1,147 @@
+// Copyright 2024 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <functional>
+#include <memory>
+#include <tuple>
+#include <type_traits>
+#include <utility>
+
+#include "ray/common/asio/instrumented_io_context.h"
+#include "ray/util/function_traits.h"
+
+namespace ray {
+
+template <typename FuncType>
+class Postable;
+
+namespace internal {
+
+template <typename FuncType>
+struct ToPostableHelper;
+
+template <typename FuncType>
+struct ToPostableHelper<std::function<FuncType>> {
+  using type = Postable<FuncType>;
+};
+
+}  // namespace internal
+
+template <typename FuncType>
+using ToPostable = typename internal::ToPostableHelper<FuncType>::type;
+
+/// Postable wraps a std::function and an instrumented_io_context together, ensuring the
+/// function can only be Post()ed or Dispatch()ed to that specific io_context. This
+/// provides type safety and prevents accidentally running the function on the wrong
+/// io_context.
+///
+/// A Postable can only be Post()ed or Dispatch()ed once. After that, it is moved-from and
+/// a next invocation will fail.
+template <typename FuncType>
+class Postable {
+ public:
+  Postable(std::function<FuncType> func, instrumented_io_context &io_context)
+      : func_(std::move(func)), io_context_(io_context) {
+    RAY_CHECK(func_ != nullptr)
```

Comment:
Just a question, I found quite a lot of our code checks `func != nullptr`, do you think worthy to have a `NotNullFunction<>`?

Reply:
Yes, it's worthwhile.

Reply:
other than in Postable I only find ~2 places to do that check. maybe we can do it later as P2.